cpp/src/arrow/python/common.cc

```diff
@@ -47,22 +48,37 @@ MemoryPool* get_memory_pool() {
 // ----------------------------------------------------------------------
 // PyBuffer
 
-PyBuffer::PyBuffer(PyObject* obj) : Buffer(nullptr, 0), obj_(nullptr) {
-  if (PyObject_CheckBuffer(obj)) {
-    obj_ = PyMemoryView_FromObject(obj);
-    Py_buffer* buffer = PyMemoryView_GET_BUFFER(obj_);
-    data_ = reinterpret_cast<const uint8_t*>(buffer->buf);
-    size_ = buffer->len;
-    capacity_ = buffer->len;
-    is_mutable_ = false;
+PyBuffer::PyBuffer() : Buffer(nullptr, 0) {}
+
+PyBuffer::PyBuffer(PyObject* obj) : Buffer(nullptr, 0) {
+  // Cannot propagate error upwards
+  (void)Init(obj);
```

Comment:
This suggests we should make the ctors private and add a static alternative constructor `Status PyBuffer::FromPyObject(PyObject* obj, std::shared_ptr<Buffer>* out)` or something. This can be declared as `@staticmethod` in Cython

Reply:
From what I can see the explicit constructor is required by the PyBytesReader constructor implementation:
```
PyBytesReader::PyBytesReader(PyObject* obj)
    : io::BufferReader(std::make_shared<PyBuffer>(obj)) {}
```

Reply:
I see. Then `PyBytesReader` will need to change, too (since making a `PyBuffer` can fail). Let me know if I can help

Reply:
Ok, done. As a side effect, this removes the `PyBytesReader` class which is not needed anymore (you can simply construct a `BufferReader` from the given `PyBuffer` pointer).

Reply:
SGTM